### DISCUSSION TITLE
Add `useRedirect` as a core composable

### DISCUSSION THREAD

**Discussion by qnp** [nuxt/nuxt#17837] [Ideas]
In the light of these discussions:
https://github.com/nuxt/framework/discussions/2082#discussioncomment-1699776
https://github.com/nuxt/framework/discussions/2701#discussioncomment-1961817
https://github.com/nuxt/framework/discussions/1722#discussioncomment-1649805
https://github.com/nuxt/framework/discussions/2004

And pointing out the fact that in Nuxt 2, one could perform a redirection after a call to `asyncData` if some condition was (or wasn’t) fulfilled directly using `redirect` from `asyncData` context:
```js
async asyncData({ redirect }) {
   // Fetch something async
   const data = await ...
   // redirect to another route under some contions
   if (/* some conditions */) return redirect('/new/path');
   ...
}
```

I would suggest adding a composable `useRedirect` in the app’s core composables (`nuxt3/dist/app` _aka_ `#app`) with this implementation:
```ts
import { RouteLocationRaw } from 'vue-router';

export function useRedirect(to: RouteLocationRaw) {
  const router = useRouter();
  const { ssrContext } = useNuxtApp();

  // Server-side redirection using h3 res from ssrContext
  if (process.server && ssrContext) {
    const { res } = ssrContext;
    const { path } = router.resolve(to);
    res.writeHead(302, { Location: path });
    res.end();
  }
  // Client-side redirection using vue-router
  else if (process.client) {
    router.beforeEach((_, __, next) => {
      return next(to);
    });
  }
}
```

One could then, in a `<script setup>`, do something very convenient like the following:
```ts
const { data, error } = await useAsyncData('name', () => /* fetch something async */);
if (/* some conditions on data or error */) useRedirect('/new/path');
```

This has been tested in our project and works perfectly.

**Comment by danielroe** [comment]
If it doesn't currently work, this should be implemented within navigateTo as a single semantic helper.

**Reply by qnp** [reply]
No, it does not currently work under `navigateTo`, because at the time of speaking [its implementation](https://github.com/nuxt/framework/blob/dbab979a2ed28b8f3d02044079f6b9039114d5ff/packages/nuxt3/src/app/composables/router.ts#L48) does not handle the server-side work.

**Reply by qnp** [reply]
I can make the PR in the `navigateTo` helper. However, how do we make sure that handling server-side redirections here using ssrContext’s h3 `res` does not introduce unwanted side effects ?

**Reply by danielroe** [reply]
Thanks! That would be perfect.

If there's an unwanted side effect, we can refactor or handle within the h3 layer. (likely a check for writableEnded after renderToString in nitro + an early return would do the trick).

**Reply by qnp** [reply]
I was digging into the code to make the PR. I'm really surprised to see this `(window as any).$nuxt.$router` to get the router instance client-side. It seems a bit hacky to rely on global object defined on window... Why won’t we use `useRouter` which returns `useNuxtApp()?.$router` and should be also available client-side ? Or am I missing something...

**Reply by danielroe** [reply]
Yes, we probably can now simplify it. 👍

**Reply by qnp** [reply]
In some cases (_e.g._ the redirect one), `navigateTo` in client-side should perform a `router.replace` instead of `router.push` so that it is truly a redirection, that does not add an entry in browser history.
What do you think of an optional boolean parameter "replace", or "redirect": `navigateTo(to: RouteLocationRaw, { replace: boolean })` ?

**Reply by danielroe** [reply]
Yes, that is a good idea. 👍 But in the case of a middleware, a router.push should not add an entry to the browser history.

**Reply by qnp** [reply]
Yes, indeed because `isProcessingMiddleware()` is true, hence `navigateTo` just returns `to` (and never reaches `router.push`), isn’t it ?

**Reply by qnp** [reply]
From my actual tests, beware that the client-side navigation does not play well with page transitions (hence, that is also the case of the current implementation)
I have to define:
```
definePageMeta({
  pageTransition: false
})
```
to avoid this error:
```
Uncaught DOMException: Failed to execute 'insertBefore' on 'Node': The node before which the new node is to be inserted is not a child of this node.
```
_Full stacktrace_
<details>
<table class="table table-striped table-bordered">
<thead>
<tr>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>insert</td>
<td>@runtime-dom.esm-bundler.js:10</td>
</tr>
<tr>
<td>move</td>
<td>@runtime-core.esm-bundler.js:5452</td>
</tr>
<tr>
<td>move</td>
<td>@runtime-core.esm-bundler.js:5422</td>
</tr>
<tr>
<td>activeBranch.transition.afterLeave</td>
<td>@runtime-core.esm-bundler.js:1367</td>
</tr>
<tr>
<td>performRemove</td>
<td>@runtime-core.esm-bundler.js:5547</td>
</tr>
<tr>
<td>el._leaveCb</td>
<td>@runtime-core.esm-bundler.js:2137</td>
</tr>
<tr>
<td>finishLeave</td>
<td>@runtime-dom.esm-bundler.js:833</td>
</tr>
<tr>
<td>resolve2</td>
<td>@runtime-dom.esm-bundler.js:863</td>
</tr>
<tr>
<td>whenTransitionEnds</td>
<td>@runtime-dom.esm-bundler.js:952</td>
</tr>
<tr>
<td>(anonymous)</td>
<td>@runtime-dom.esm-bundler.js:872</td>
</tr>
<tr>
<td>requestAnimationFrame (async)</td>
<td></td>
</tr>
<tr>
<td>(anonymous)</td>
<td>@runtime-dom.esm-bundler.js:936</td>
</tr>
<tr>
<td>requestAnimationFrame (async)</td>
<td></td>
</tr>
<tr>
<td>nextFrame</td>
<td>@runtime-dom.esm-bundler.js:935</td>
</tr>
<tr>
<td>onLeave</td>
<td>@runtime-dom.esm-bundler.js:868</td>
</tr>
<tr>
<td>leave</td>
<td>@runtime-core.esm-bundler.js:2151</td>
</tr>
<tr>
<td>performLeave</td>
<td>@runtime-core.esm-bundler.js:5554</td>
</tr>
<tr>
<td>remove2</td>
<td>@runtime-core.esm-bundler.js:5559</td>
</tr>
<tr>
<td>unmount</td>
<td>@runtime-core.esm-bundler.js:5521</td>
</tr>
<tr>
<td>unmountComponent</td>
<td>@runtime-core.esm-bundler.js:5593</td>
</tr>
<tr>
<td>unmount</td>
<td>@runtime-core.esm-bundler.js:5494</td>
</tr>
<tr>
<td>resolve</td>
<td>@runtime-core.esm-bundler.js:1378</td>
</tr>
<tr>
<td>patchSuspense</td>
<td>@runtime-core.esm-bundler.js:1302</td>
</tr>
<tr>
<td>process</td>
<td>@runtime-core.esm-bundler.js:1181</td>
</tr>
<tr>
<td>patch</td>
<td>@runtime-core.esm-bundler.js:4495</td>
</tr>
<tr>
<td>componentUpdateFn</td>
<td>@runtime-core.esm-bundler.js:5107</td>
</tr>
<tr>
<td>run</td>
<td>@reactivity.esm-bundler.js:167</td>
</tr>
<tr>
<td>updateComponent</td>
<td>@runtime-core.esm-bundler.js:4968</td>
</tr>
<tr>
<td>processComponent</td>
<td>@runtime-core.esm-bundler.js:4901</td>
</tr>
<tr>
<td>patch</td>
<td>@runtime-core.esm-bundler.js:4489</td>
</tr>
<tr>
<td>componentUpdateFn</td>
<td>@runtime-core.esm-bundler.js:5107</td>
</tr>
<tr>
<td>run</td>
<td>@reactivity.esm-bundler.js:167</td>
</tr>
<tr>
<td>updateComponent</td>
<td>@runtime-core.esm-bundler.js:4968</td>
</tr>
<tr>
<td>processComponent</td>
<td>@runtime-core.esm-bundler.js:4901</td>
</tr>
<tr>
<td>patch</td>
<td>@runtime-core.esm-bundler.js:4489</td>
</tr>
<tr>
<td>componentUpdateFn</td>
<td>@runtime-core.esm-bundler.js:5107</td>
</tr>
<tr>
<td>run</td>
<td>@reactivity.esm-bundler.js:167</td>
</tr>
<tr>
<td>callWithErrorHandling</td>
<td>@runtime-core.esm-bundler.js:155</td>
</tr>
<tr>
<td>flushJobs</td>
<td>@runtime-core.esm-bundler.js:394</td>
</tr>
<tr>
<td>Promise.then (async)</td>
<td></td>
</tr>
<tr>
<td>queueFlush</td>
<td>@runtime-core.esm-bundler.js:285</td>
</tr>
<tr>
<td>queueJob</td>
<td>@runtime-core.esm-bundler.js:279</td>
</tr>
<tr>
<td>(anonymous)</td>
<td>@runtime-core.esm-bundler.js:5139</td>
</tr>
<tr>
<td>triggerEffects</td>
<td>@reactivity.esm-bundler.js:359</td>
</tr>
<tr>
<td>triggerRefValue</td>
<td>@reactivity.esm-bundler.js:957</td>
</tr>
<tr>
<td>(anonymous)</td>
<td>@reactivity.esm-bundler.js:1088</td>
</tr>
<tr>
<td>triggerEffects</td>
<td>@reactivity.esm-bundler.js:359</td>
</tr>
<tr>
<td>triggerRefValue</td>
<td>@reactivity.esm-bundler.js:957</td>
</tr>
<tr>
<td>set value</td>
<td>@reactivity.esm-bundler.js:1001</td>
</tr>
<tr>
<td>finalizeNavigation</td>
<td>@vue-router.esm-bundler.js:3194</td>
</tr>
<tr>
<td>(anonymous)</td>
<td>@vue-router.esm-bundler.js:3067</td>
</tr>
<tr>
<td>Promise.then (async)</td>
<td></td>
</tr>
<tr>
<td>pushWithRedirect</td>
<td>@vue-router.esm-bundler.js:3038</td>
</tr>
<tr>
<td>push</td>
<td>@vue-router.esm-bundler.js:2966</td>
</tr>
<tr>
<td>replace</td>
<td>@vue-router.esm-bundler.js:2969</td>
</tr>
<tr>
<td>navigateTo</td>
<td>@router.ts:66</td>
</tr>
<tr>
<td>setup</td>
<td>@navigate-to.vue:7</td>
</tr>
<tr>
<td>await in setup (async)</td>
<td></td>
</tr>
<tr>
<td>callWithErrorHandling</td>
<td>@runtime-core.esm-bundler.js:155</td>
</tr>
<tr>
<td>setupStatefulComponent</td>
<td>@runtime-core.esm-bundler.js:7056</td>
</tr>
<tr>
<td>setupComponent</td>
<td>@runtime-core.esm-bundler.js:7012</td>
</tr>
<tr>
<td>mountComponent</td>
<td>@runtime-core.esm-bundler.js:4922</td>
</tr>
<tr>
<td>processComponent</td>
<td>@runtime-core.esm-bundler.js:4897</td>
</tr>
<tr>
<td>patch</td>
<td>@runtime-core.esm-bundler.js:4489</td>
</tr>
<tr>
<td>patchSuspense</td>
<td>@runtime-core.esm-bundler.js:1299</td>
</tr>
<tr>
<td>process</td>
<td>@runtime-core.esm-bundler.js:1181</td>
</tr>
<tr>
<td>patch</td>
<td>@runtime-core.esm-bundler.js:4495</td>
</tr>
<tr>
<td>componentUpdateFn</td>
<td>@runtime-core.esm-bundler.js:5107</td>
</tr>
<tr>
<td>run</td>
<td>@reactivity.esm-bundler.js:167</td>
</tr>
<tr>
<td>updateComponent</td>
<td>@runtime-core.esm-bundler.js:4968</td>
</tr>
<tr>
<td>processComponent</td>
<td>@runtime-core.esm-bundler.js:4901</td>
</tr>
<tr>
<td>patch</td>
<td>@runtime-core.esm-bundler.js:4489</td>
</tr>
<tr>
<td>componentUpdateFn</td>
<td>@runtime-core.esm-bundler.js:5107</td>
</tr>
<tr>
<td>run</td>
<td>@reactivity.esm-bundler.js:167</td>
</tr>
<tr>
<td>updateComponent</td>
<td>@runtime-core.esm-bundler.js:4968</td>
</tr>
<tr>
<td>processComponent</td>
<td>@runtime-core.esm-bundler.js:4901</td>
</tr>
<tr>
<td>patch</td>
<td>@runtime-core.esm-bundler.js:4489</td>
</tr>
<tr>
<td>componentUpdateFn</td>
<td>@runtime-core.esm-bundler.js:5107</td>
</tr>
<tr>
<td>run</td>
<td>@reactivity.esm-bundler.js:167</td>
</tr>
<tr>
<td>callWithErrorHandling</td>
<td>@runtime-core.esm-bundler.js:155</td>
</tr>
<tr>
<td>flushJobs</td>
<td>@runtime-core.esm-bundler.js:394</td>
</tr>
<tr>
<td>Promise.then (async)</td>
<td></td>
</tr>
<tr>
<td>queueFlush</td>
<td>@runtime-core.esm-bundler.js:285</td>
</tr>
<tr>
<td>queueJob</td>
<td>@runtime-core.esm-bundler.js:279</td>
</tr>
<tr>
<td>(anonymous)</td>
<td>@runtime-core.esm-bundler.js:5139</td>
</tr>
<tr>
<td>triggerEffects</td>
<td>@reactivity.esm-bundler.js:359</td>
</tr>
<tr>
<td>triggerRefValue</td>
<td>@reactivity.esm-bundler.js:957</td>
</tr>
<tr>
<td>(anonymous)</td>
<td>@reactivity.esm-bundler.js:1088</td>
</tr>
<tr>
<td>triggerEffects</td>
<td>@reactivity.esm-bundler.js:359</td>
</tr>
<tr>
<td>triggerRefValue</td>
<td>@reactivity.esm-bundler.js:957</td>
</tr>
<tr>
<td>set value</td>
<td>@reactivity.esm-bundler.js:1001</td>
</tr>
<tr>
<td>finalizeNavigation</td>
<td>@vue-router.esm-bundler.js:3194</td>
</tr>
<tr>
<td>(anonymous)</td>
<td>@vue-router.esm-bundler.js:3067</td>
</tr>
<tr>
<td>Promise.then (async)</td>
<td></td>
</tr>
<tr>
<td>pushWithRedirect</td>
<td>@vue-router.esm-bundler.js:3038</td>
</tr>
<tr>
<td>push</td>
<td>@vue-router.esm-bundler.js:2966</td>
</tr>
<tr>
<td>navigate</td>
<td>@vue-router.esm-bundler.js:2093</td>
</tr>
<tr>
<td>callWithErrorHandling</td>
<td>@runtime-core.esm-bundler.js:155</td>
</tr>
<tr>
<td>callWithAsyncErrorHandling</td>
<td>@runtime-core.esm-bundler.js:164</td>
</tr>
<tr>
<td>invoker</td>
<td>@runtime-dom.esm-bundler.js:366</td>
</tr>
</tbody>
</table>
</details>

**Reply by danielroe** [reply]
Yes, that may be the same upstream bug in vue itself as https://github.com/vuejs/router/issues/1319 and https://github.com/nuxt/framework/issues/3141.

**Reply by qnp** [reply]
PR done: #3684
